modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,200 @@
+    Ref.of[F, PosLong](PosLong.MinValue),
+    Ref.of[F, NonNegLong](NonNegLong.MinValue),
+    Ref.of[F, Boolean](false)
+  ).mapN(make[F](_, _, _, _, identifierStorage, globalL0ClusterStorage, lastGlobalSnapshotStorage, stateChannelSnapshotClient))
+
+  def make[F[_]: Async: Hasher: Temporal](
+    pendingR: Ref[F, Queue[TrackedBinary]],
+    expectedOrdinalsToConfirmR: Ref[F, PosLong],
+    avgConfirmedCountPerSnapshotR: Ref[F, NonNegLong],
+    catchUpMode: Ref[F, Boolean],
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): StateChannelBinarySender[F] =
+    new StateChannelBinarySender[F] {
+
+      private val logger = Slf4jLogger.getLogger
+
+      def process(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        catchUpMode.get.ifM(
+          enqueue(binary),
+          enqueue(binary) >> post(binary)
+        )
+
+      def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit] = for {
+        identifier <- identifierStorage.get
+        binaries = globalSnapshot.stateChannelSnapshots.get(identifier).toList.flatMap(_.toList)
+        confirmedHashesInGlobalSnapshot <- binaries.traverse(_.toHashed).map(_.flatMap(b => List(b.hash, b.lastSnapshotHash)))
+
+        (confirmed, pending) <- pendingR.modify { prevPending =>
+          val (confirmedWithIndex, unconfirmedWithIndex) = prevPending.zipWithIndex.partition {
+            case (TrackedBinary(tracked, _), _) => confirmedHashesInGlobalSnapshot.contains(tracked.hash)
+          }
+
+          val unconfirmed = unconfirmedWithIndex.map { case (binary, _) => binary }
+          val confirmed = confirmedWithIndex.map { case (binary, _) => binary }
+          val pending = confirmedWithIndex.lastOption.map { case (_, index) => index }
+            .map(cuttingIndex => unconfirmedWithIndex.collect { case (binary, index) if index >= cuttingIndex => binary })
+            .getOrElse(unconfirmed)
+
+          (pending, (confirmed, pending))
+        }
+
+        expectedOrdinalsToConfirm <- updateExpectedOrdinalsToConfirm(confirmed, globalSnapshot.ordinal)
+        _ <- updateAvgConfirmedCountPerSnapshot(confirmed)
+        _ <- updateCatchUpMode(pending, globalSnapshot.ordinal, expectedOrdinalsToConfirm)
+
+        _ <- logger
+          .warn(
+            s"${pending.length} state channel binaries are still not confirmed on the global state. Waiting for the next global snapshot."
+          )
+          .whenA(pending.nonEmpty)
+      } yield ()
+
+      def processPending: F[Unit] = {
+        val retry = for {
+          pending <- pendingR.get
+          avgConfirmedPerSnapshot <- avgConfirmedCountPerSnapshotR.get
+          cap = Math
+            .round(avgConfirmedPerSnapshot.toDouble + (avgConfirmedPerSnapshot.toDouble / 2))
```

Comment:
This rounds to 0 for values smaller than 0.4. So for such a case we will stop retrying at all. The question is if that's unlikely to go down to such a small number or not (maybe it's ok to let it just get stuck eventually). I would actually add some kind of retry counter counting "number of retries without new confirmations". Once the average is small like 0.2 (and decreasing) and counter increase we should retry at least once every N retries if we multiply average by retry counter (increasing counter would need to have greater impact than decreasing average).

Reply:
Yeah. Currently, when there are no confirmed snapshots, the average confirmations count gets divided by 2 which means that the value is going to go to zero. The quick and simple solution would be to use `ceil` so there is always at least `1` to retry. On the other hand, if Global L0 is stuck for some reason - we do not want to keep spamming with snapshots.

Reply:
>if we multiply average by retry counter

So what you propose is the following:
For avg = 10 and retriesWithoutConfirmation = 3 we should retry at least 1 snapshot after 30 ordinals?

Reply:
How about this:

```scala
val ordinalsWithoutConfirmation = ...
val maxOrdinalsWithoutConfirmation = 5

val cap = if (ordinalsWithoutConfirmation >= maxOrdinalsWithoutConfirmation && average < 1) {
    if (ordinalsWithoutConfirmation % 5 == 0) 1 else 0  // If average < 1 try to send 1 every 5 ordinals
} else {
   average
}
```